journalbeat/reader/journal.go

```diff
@@ -0,0 +1,258 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package reader
+
+import (
+	"io"
+	"os"
+	"time"
+
+	"github.com/coreos/go-systemd/sdjournal"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/journalbeat/checkpoint"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+const (
+	// LocalSystemJournalID is the ID of the local system journal.
+	LocalSystemJournalID = "LOCAL_SYSTEM_JOURNAL"
+)
+
+// Config stores the options of a reder.
+type Config struct {
+	// Path is the path to the journal file.
+	Path string
+	// Seek specifies the seeking stategy.
+	// Possible values: head, tail, cursor.
+	Seek string
+	// MaxBackoff is the limit of the backoff time.
+	MaxBackoff time.Duration
+	// Backoff is the current interval to wait before
+	// attemting to read again from the journal.
+	Backoff time.Duration
+	// BackoffFactor is the multiplier of Backoff.
+	BackoffFactor int
+}
+
+// Reader reads entries from journal(s).
+type Reader struct {
+	j       *sdjournal.Journal
+	config  Config
+	changes chan int
+	done    chan struct{}
+}
+
+// New creates a new journal reader and moves the FP to the configured position.
```

Comment:
I think this type would benefit from having a scoped logger, it will be useful to know which input created that logger when we are debugging users errors.